core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/WorkflowPipelinedRegions.scala

```diff
@@ -0,0 +1,177 @@
+          val prevInOrder = getPipelinedRegionFromOperatorId(
+            toOperatorIdentity(inputProcessingOrderForOp(i - 1).from)
+          )
+          val nextInOrder =
+            getPipelinedRegionFromOperatorId(toOperatorIdentity(inputProcessingOrderForOp(i).from))
+
+          if (
+            prevInOrder.getId() != nextInOrder.getId() && (!regionDependence.contains(
+              nextInOrder
+            ) || !regionDependence(nextInOrder).contains(prevInOrder.getId()))
+          ) {
+            val dependencies = regionDependence.getOrElseUpdate(
+              nextInOrder,
+              new ArrayBuffer[PipelinedRegionIdentity]()
+            )
+            dependencies.append(prevInOrder.getId())
+            regionDependence(nextInOrder) = dependencies
+          }
+        }
+      }
+
+      // 2. Find dependencies due to blocking operators.
+      // e.g. The region before and after a sort operator has dependencies
+      val upstreamOps = workflow.getDirectUpstreamOperators(opId)
+      upstreamOps.foreach(upstreamOp => {
+        val linkFromUpstreamOp = LinkIdentity(
+          workflow.getOperator(upstreamOp).topology.layers.last.id,
+          workflow.getOperator(opId).topology.layers.head.id
+        )
+        if (workflow.getOperator(opId).isInputBlocking(linkFromUpstreamOp)) {
+          val prevInOrder = getPipelinedRegionFromOperatorId(upstreamOp)
+          val nextInOrder = getPipelinedRegionFromOperatorId(opId)
+          if (prevInOrder.getId() != nextInOrder.getId()) {
+            if (
+              !regionDependence.contains(
+                nextInOrder
+              ) || !regionDependence(nextInOrder).contains(prevInOrder.getId())
+            ) {
+              val dependencies = regionDependence.getOrElseUpdate(
+                nextInOrder,
+                new ArrayBuffer[PipelinedRegionIdentity]()
+              )
+              dependencies.append(prevInOrder.getId())
+              regionDependence(nextInOrder) = dependencies
+            }
+
+            if (
+              !regionTerminalOperatorInOtherRegions.contains(
+                prevInOrder
+              ) || !regionTerminalOperatorInOtherRegions(prevInOrder).contains(opId)
+            ) {
+
+              val terminalOps = regionTerminalOperatorInOtherRegions.getOrElseUpdate(
+                prevInOrder,
+                new ArrayBuffer[OperatorIdentity]()
+              )
+              terminalOps.append(opId)
+              regionTerminalOperatorInOtherRegions(prevInOrder) = terminalOps
+            }
+
+          }
+        }
+      })
+    })
+
+    for ((region, dependence) <- regionDependence) {
+      region.dependsOn = dependence.toArray
+    }
+    for ((region, terminalOps) <- regionTerminalOperatorInOtherRegions) {
+      region.blockingDowstreamOperatorsInOtherRegions = terminalOps.toArray
+    }
+  }
+
+  private def createPipelinedRegions(): Unit = {
+    findAllPipelinedRegions()
+    orderRegions()
```

Comment:
We can use jgrapht library to implement algorithm to find weakly connected components and topological ordering stuff:

Check the following:
the `WorkflowCompiler` is already using a variable called `jgraphtDag`

https://jgrapht.org/javadoc-1.4.0/org/jgrapht/alg/connectivity/BiconnectivityInspector.html
https://jgrapht.org/javadoc/org.jgrapht.core/org/jgrapht/traverse/TopologicalOrderIterator.html

Reply:
done